drivers/ethernet/phy_xlnx_gem.c

```diff
@@ -0,0 +1,975 @@
+	 * Page 0, register address 1 =
+	 *     Copper Status Register, bit [5] = Copper Auto-Negotiation
+	 *     Complete.
+	 *
+	 * Comp. datasheet tables 68 & 73.
+	 */
+
+	/*
+	 * 88E151x only:
+	 * Register 4, bits [4..0] = Selector field, 00001 = 802.3. Those bits
+	 * are reserved in other Marvell PHYs.
+	 */
+	if ((dev_data->phy_id & PHY_MRVL_PHY_ID_MODEL_MASK) ==
+		PHY_MRVL_PHY_ID_MODEL_88E151X) {
+		phy_data = PHY_MRVL_ADV_SELECTOR_802_3;
+	} else {
+		phy_data = 0x0000;
+	}
+
+	/*
+	 * Clear the 1 GBit/s FDX/HDX advertisement bits from reg. 9's current
+	 * contents in case we're going to advertise anything below 1 GBit/s
+	 * as maximum / nominal link speed.
+	 */
+	phy_data_gbit = phy_xlnx_gem_mdio_read(dev_conf->base_addr,
+		dev_data->phy_addr, PHY_MRVL_1000BASET_CONTROL_REGISTER);
+	phy_data_gbit &= ~(PHY_MRVL_ADV_1000BASET_FDX_BIT |
+		PHY_MRVL_ADV_1000BASET_HDX_BIT);
+
+	if (dev_conf->enable_fdx) {
+		if (dev_conf->max_link_speed == LINK_1GBIT) {
+			/* Advertise 1 GBit/s, full duplex */
+			phy_data_gbit |= PHY_MRVL_ADV_1000BASET_FDX_BIT;
+			if (dev_conf->phy_advertise_lower) {
+				/* + 100 MBit/s, full duplex */
+				phy_data |= PHY_MRVL_ADV_100BASET_FDX_BIT;
+				/* + 10 MBit/s, full duplex */
+				phy_data |= PHY_MRVL_ADV_10BASET_FDX_BIT;
+			}
+		} else if (dev_conf->max_link_speed == LINK_100MBIT) {
+			/* Advertise 100 MBit/s, full duplex */
+			phy_data |= PHY_MRVL_ADV_100BASET_FDX_BIT;
+			if (dev_conf->phy_advertise_lower) {
+				/* + 10 MBit/s, full duplex */
+				phy_data |= PHY_MRVL_ADV_10BASET_FDX_BIT;
+			}
+		} else if (dev_conf->max_link_speed == LINK_10MBIT) {
+			/* Advertise 10 MBit/s, full duplex */
+			phy_data |= PHY_MRVL_ADV_10BASET_FDX_BIT;
+		}
+	} else {
+		if (dev_conf->max_link_speed == LINK_1GBIT) {
+			/* Advertise 1 GBit/s, half duplex */
+			phy_data_gbit = PHY_MRVL_ADV_1000BASET_HDX_BIT;
+			if (dev_conf->phy_advertise_lower) {
+				/* + 100 MBit/s, half duplex */
+				phy_data |= PHY_MRVL_ADV_100BASET_HDX_BIT;
+				/* + 10 MBit/s, half duplex */
+				phy_data |= PHY_MRVL_ADV_10BASET_HDX_BIT;
+			}
+		} else if (dev_conf->max_link_speed == LINK_100MBIT) {
+			/* Advertise 100 MBit/s, half duplex */
+			phy_data |= PHY_MRVL_ADV_100BASET_HDX_BIT;
+			if (dev_conf->phy_advertise_lower) {
+				/* + 10 MBit/s, half duplex */
+				phy_data |= PHY_MRVL_ADV_10BASET_HDX_BIT;
+			}
+		} else if (dev_conf->max_link_speed == LINK_10MBIT) {
+			/* Advertise 10 MBit/s, half duplex */
+			phy_data |= PHY_MRVL_ADV_10BASET_HDX_BIT;
+		}
+	}
+
+	phy_xlnx_gem_mdio_write(dev_conf->base_addr, dev_data->phy_addr,
+		PHY_MRVL_1000BASET_CONTROL_REGISTER, phy_data_gbit);
+	phy_xlnx_gem_mdio_write(dev_conf->base_addr, dev_data->phy_addr,
+		PHY_MRVL_COPPER_AUTONEG_ADV_REGISTER, phy_data);
+
+	/*
+	 * Page 0 / Reg 0 [15] Copper Software Reset bit, affecting pages
+	 * 0, 2, 3, 5, 7. Reset is performed immediately, bit [15] is self-
+	 * clearing.
+	 */
+	phy_data  = phy_xlnx_gem_mdio_read(dev_conf->base_addr,
```

Comment:
same comment about the reset.  maybe pass autoneg as a parameter to the reset function?